__1__. Импортируйте все необходимые библиотеки.

In [113]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', 100)
pd.set_option('display.precision', 3)

import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

__2__. Загрузите файл `train.csv` и сохраните его в переменную `df`. Выведите первые 10 строчек таблицы.

In [114]:
df = pd.read_csv('train.csv')
df.head(10)

PERIOD  cl_id   MCC channel_type  currency        TRDATETIME   amount  \
0  01/10/2017      0  5200          NaN       810  21OCT17:00:00:00   5023.0   
1  01/10/2017      0  6011          NaN       810  12OCT17:12:24:07  20000.0   
2  01/12/2017      0  5921          NaN       810  05DEC17:00:00:00    767.0   
3  01/10/2017      0  5411          NaN       810  21OCT17:00:00:00   2031.0   
4  01/10/2017      0  6012          NaN       810  24OCT17:13:14:24  36562.0   
5  01/10/2017      1  5814          NaN       810  16OCT17:00:00:00    380.0   
6  01/10/2017      1  5814          NaN       810  10OCT17:00:00:00    378.0   
7  01/10/2017      1  5814          NaN       810  16OCT17:00:00:00    199.0   
8  01/10/2017      1  5814          NaN       810  11OCT17:00:00:00    400.0   
9  01/07/2017      1  5411          NaN       810  26JUL17:00:00:00    598.0   

  trx_category  target_flag  target_sum  
0          POS            0         0.0  
1      DEPOSIT            0         0.0  
2          POS            0         0.0  
3          POS            0         0.0  
4      C2C_OUT            0         0.0  
5          POS            0         0.0  
6          POS            0         0.0  
7          POS            0         0.0  
8          POS            0         0.0  
9          POS            0         0.0

In [115]:
np.random.seed(9564)
index = np.random.choice(len(df), replace=False, size=1000)

__3__. Выведите размер получившейся таблицы. Выведите статистики (среднее, дисперссию, кванти́ли, количество значений и т.п.) по каждому из столбцов. Для последнего действия воспользуйтесь специальным методом у `df`.

In [116]:
df.shape

(490513, 10)

In [117]:
df.describe()

cl_id         MCC    currency     amount  target_flag  target_sum
count  490513.000  490513.000  490513.000  4.905e+05   490513.000   4.905e+05
mean     4802.127    5699.784     810.602  5.724e+03        0.653   1.627e+05
std      2986.493     662.933      38.223  7.704e+04        0.476   1.474e+06
min         0.000     742.000      32.000  4.000e-02        0.000   0.000e+00
25%      2193.000    5411.000     810.000  2.240e+02        0.000   0.000e+00
50%      4576.000    5661.000     810.000  6.240e+02        1.000   2.491e+04
75%      7390.000    5941.000     810.000  2.176e+03        1.000   1.516e+05
max     10215.000    9405.000     986.000  3.584e+07        1.000   4.607e+07

Опиание столбцов:
* PERIOD – месяц транзакции; 
* cl_id – id клиента;
* MCC – код категории продавца;
* channel_type – канал привлечения клиента;
* currency – валюта;
* TRDATETIME – дата/время транзакции;
* amount – сумма транзакции;
* trx_category – вид транзакции.

Виды транзакций:
* POS – оплата через POS терминал;
* C2C_IN – перевод на карту (входящий платёж);
* C2C_OUT – перевод на карту (исходящий платёж);
* DEPOSIT – пополнение карты в банкомате;
* WD_ATM_PARTNER – снятие наличных в банкоматах партнерах банка *UNKNOWN*;
* WD_ATM_ROS – снятие наличных в банкоматах банка *UNKNOWN*;
* WD_ATM_OTHER – снятия в других банкоматах;
* BACK_TRX – возврат средств при возврате покупки;
* CAT – операции в банкоматах;
* CASH_ADV – снятие карты в кассе банка.

__4__. Какие столбцы имеют пропущенные значения? Сколько пропущенных значений?

In [118]:
df.info()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 490513 entries, 0 to 490512
Data columns (total 10 columns):
PERIOD          490513 non-null object
cl_id           490513 non-null int64
MCC             490513 non-null int64
channel_type    487603 non-null object
currency        490513 non-null int64
TRDATETIME      490513 non-null object
amount          490513 non-null float64
trx_category    490513 non-null object
target_flag     490513 non-null int64
target_sum      490513 non-null float64
dtypes: float64(2), int64(4), object(4)
memory usage: 37.4+ MB


PERIOD             0
cl_id              0
MCC                0
channel_type    2910
currency           0
TRDATETIME         0
amount             0
trx_category       0
target_flag        0
target_sum         0
dtype: int64

Столбец channel_type имеет 2910 пропущенных значений

__5__. Какие уникальные значения существуют у столбца `currency`?

In [119]:
df["currency"].unique()

array([810, 978, 504, 704, 981, 985, 840, 949,  51, 826, 214, 764, 203,
       702, 360, 756, 933, 975,  36, 191, 784, 980, 124, 398, 376, 944,
       352, 417, 156, 752, 392, 484, 634, 188, 643, 348, 356, 458, 986,
       498, 578, 208, 344,  32, 410, 788, 480, 604, 941, 144, 946, 710,
       690,  44, 170, 901, 608, 554, 462])

__6__. Заполните пропуски в столбце `channel_type` значением `type0`. Результат должен сохраниться в переменной `df`.

In [120]:
df.loc[df['channel_type'].isna(), 'channel_type'] = 'type1'

In [121]:
df.head()

PERIOD  cl_id   MCC channel_type  currency        TRDATETIME   amount  \
0  01/10/2017      0  5200        type1       810  21OCT17:00:00:00   5023.0   
1  01/10/2017      0  6011        type1       810  12OCT17:12:24:07  20000.0   
2  01/12/2017      0  5921        type1       810  05DEC17:00:00:00    767.0   
3  01/10/2017      0  5411        type1       810  21OCT17:00:00:00   2031.0   
4  01/10/2017      0  6012        type1       810  24OCT17:13:14:24  36562.0   

  trx_category  target_flag  target_sum  
0          POS            0         0.0  
1      DEPOSIT            0         0.0  
2          POS            0         0.0  
3          POS            0         0.0  
4      C2C_OUT            0         0.0

__7__. Посчитайте, сколько записей приходится на каждое уникальное значение столбца `channel_type`.

In [122]:
df['channel_type'].unique()
print(df.loc[df['channel_type'] == 'type1' , 'channel_type'].shape)
print(df.loc[df['channel_type'] == 'type2' , 'channel_type'].shape)
print(df.loc[df['channel_type'] == 'type3' , 'channel_type'].shape)
print(df.loc[df['channel_type'] == 'type4' , 'channel_type'].shape)
print(df.loc[df['channel_type'] == 'type5' , 'channel_type'].shape)

(302157,)
(175013,)
(3280,)
(4476,)
(5587,)


__8__. Найдите топ-3 `MCC` категорий с наибольшим объемом трат. 

Сгруппируем по МСС и найдем максимальное по сумме  amount

In [123]:
df_MCC_amount = df[['MCC', 'amount']].groupby('MCC').sum().reset_index()
df_MCC_amount.head()
df_MCC_amount.nlargest(3, 'amount')

MCC     amount
244  6011  1.494e+09
245  6012  1.849e+08
158  5411  1.347e+08

__9__. Сделайте значения в столбцах `channel_type` и `trx_category` категориальными.

In [124]:
df[['channel_type', 'trx_category']] = \
    df[['channel_type', 'trx_category']].astype('category')

df['channel_type_label'] = df['channel_type'].cat.codes
df['trx_category_label'] = df['trx_category'].cat.codes
#np.random.seed(2382)
#df[['trx_category', 'trx_category_label']].loc[np.random.randint(0, df.shape[0], 10)]

__10__. Сделайте столбец `trx_category_mul`, которая показывает увеличение (значение в столбце равно +1) и уменьшение средств (значение в столбце равно -1) на счете клиента банка. Операции, которые увеличивают количество средств на счете:
`C2C_IN`, `DEPOSIT`, `BACK_TRX`. Считайте, что остальные операции уменьшают количество средств на счете.

In [125]:
def oper(trx):
    if trx == 'C2C_IN' or trx == 'DEPOSIT' or trx == 'BACK_TRX':
        trx_m = 1
    else:
        trx_m = -1
    return trx_m
df.loc[:, 'trx_category_mul'] = df.loc[:, 'trx_category'].apply(oper)

In [126]:
df.loc[:, ['trx_category_mul', 'trx_category']].head()

trx_category_mul trx_category
0                -1          POS
1                 1      DEPOSIT
2                -1          POS
3                -1          POS
4                -1      C2C_OUT

__11__. Приведем столбец `TRDATETIME` к типу даты.

In [127]:
from datetime import datetime
from datetime import date

df["TRDATETIME"] = pd.to_datetime(df["TRDATETIME"], format='%d%b%y:%H:%M:%S')
df.head()

PERIOD  cl_id   MCC channel_type  currency          TRDATETIME  \
0  01/10/2017      0  5200        type1       810 2017-10-21 00:00:00   
1  01/10/2017      0  6011        type1       810 2017-10-12 12:24:07   
2  01/12/2017      0  5921        type1       810 2017-12-05 00:00:00   
3  01/10/2017      0  5411        type1       810 2017-10-21 00:00:00   
4  01/10/2017      0  6012        type1       810 2017-10-24 13:14:24   

    amount trx_category  target_flag  target_sum  channel_type_label  \
0   5023.0          POS            0         0.0                   0   
1  20000.0      DEPOSIT            0         0.0                   0   
2    767.0          POS            0         0.0                   0   
3   2031.0          POS            0         0.0                   0   
4  36562.0      C2C_OUT            0         0.0                   0   

   trx_category_label  trx_category_mul  
0                   6                -1  
1                   5                 1  
2                   6                -1  
3                   6                -1  
4                   2                -1

Извлеките из столбца `TRDATETIME` следующие признаки:
* `mday` – день (с начала месяца);
* `ymonth` – месяц;
* `year` – год;
* `wday` – день недели;
* `is_weekend` – является ли день выходным днем.

Извлеченные признаки запишите в таблицу `df_time`.

In [128]:
df_time = pd.DataFrame(np.zeros((490513, 0)))
df_time['mday'] = df["TRDATETIME"].dt.day
df_time['year'] = df["TRDATETIME"].dt.year
df_time['ymonth'] = df["TRDATETIME"].dt.month
df_time['wday'] = df['TRDATETIME'].dt.weekday

def func_(wday):
    if wday < 5:
        return False
    else:
        return True
df_time.loc[:, 'is_weekend'] = df_time.loc[:,'wday'].apply(func_)


df_time.head()

mday  year  ymonth  wday  is_weekend
0    21  2017      10     5        True
1    12  2017      10     3       False
2     5  2017      12     1       False
3    21  2017      10     5        True
4    24  2017      10     1       False

__12__. Объедините таблицы `df` и `df_time` и сохраните результат в `df`. При объединении используйте аргумент `sort=False`, если он имеется.

In [129]:
df = pd.concat([df, df_time], sort=False, axis=1)

In [130]:
df.head()

PERIOD  cl_id   MCC channel_type  currency          TRDATETIME  \
0  01/10/2017      0  5200        type1       810 2017-10-21 00:00:00   
1  01/10/2017      0  6011        type1       810 2017-10-12 12:24:07   
2  01/12/2017      0  5921        type1       810 2017-12-05 00:00:00   
3  01/10/2017      0  5411        type1       810 2017-10-21 00:00:00   
4  01/10/2017      0  6012        type1       810 2017-10-24 13:14:24   

    amount trx_category  target_flag  target_sum  channel_type_label  \
0   5023.0          POS            0         0.0                   0   
1  20000.0      DEPOSIT            0         0.0                   0   
2    767.0          POS            0         0.0                   0   
3   2031.0          POS            0         0.0                   0   
4  36562.0      C2C_OUT            0         0.0                   0   

   trx_category_label  trx_category_mul  mday  year  ymonth  wday  is_weekend  
0                   6                -1    21  2017      10     5        True  
1                   5                 1    12  2017      10     3       False  
2                   6                -1     5  2017      12     1       False  
3                   6                -1    21  2017      10     5        True  
4                   2                -1    24  2017      10     1       False

__13__. Загрузите таблицу `mcc_codes_ru.csv` и сохраните в переменную `df_mcc`. Выведите первые 5 строчек таблицы.

In [131]:
df_mcc = pd.read_csv('mcc_codes_ru.csv')
df_mcc.head(5)

MCC                                           Название  \
0   742                                Ветеринарные услуги   
1   763                   Сельскохозяйственные кооперативы   
2   780          Услуги садоводства и ландшафтного дизайна   
3  1520  Генеральные подрядчики – жилое и коммерческое ...   
4  1711  Генеральные подрядчики по вентиляции, теплосна...   

               Группа Обновлено  
0  Контрактные услуги  31.01.18  
1  Контрактные услуги  31.01.18  
2  Контрактные услуги  31.01.18  
3  Контрактные услуги  31.01.18  
4  Контрактные услуги  31.01.18

__14__. Переименуйте столбцы в `df_mcc` следующим образом:
* `MCC` -> `MCC`;
* `Название` -> `mcc_name`;
* `Группа` -> `mcc_group`;
* `Обновлено` -> `updated`.

Результат сохраните в `df_mcc`.

In [132]:
df_mcc.rename(columns={'MCC': 'MCC', 'Название': 'mcc_name', 'Группа': 'mcc_group', 'Обновлено': 'updated'}, inplace=True)
df_mcc.head()

MCC                                           mcc_name  \
0   742                                Ветеринарные услуги   
1   763                   Сельскохозяйственные кооперативы   
2   780          Услуги садоводства и ландшафтного дизайна   
3  1520  Генеральные подрядчики – жилое и коммерческое ...   
4  1711  Генеральные подрядчики по вентиляции, теплосна...   

            mcc_group   updated  
0  Контрактные услуги  31.01.18  
1  Контрактные услуги  31.01.18  
2  Контрактные услуги  31.01.18  
3  Контрактные услуги  31.01.18  
4  Контрактные услуги  31.01.18

__15__. Удалите столбец `updated` из `df_mcc`, результат сохраните в `df_mcc`.

In [133]:
df_mcc = df_mcc.drop(['updated'], axis=1)
df_mcc.head()

MCC                                           mcc_name           mcc_group
0   742                                Ветеринарные услуги  Контрактные услуги
1   763                   Сельскохозяйственные кооперативы  Контрактные услуги
2   780          Услуги садоводства и ландшафтного дизайна  Контрактные услуги
3  1520  Генеральные подрядчики – жилое и коммерческое ...  Контрактные услуги
4  1711  Генеральные подрядчики по вентиляции, теплосна...  Контрактные услуги

__16__. Закодируйте столбец `mcc_group` с помощью label-кодирования.

In [134]:
df_mcc[['mcc_group']] = df_mcc[['mcc_group']].astype('category')
df_mcc['mcc_group'] = df_mcc['mcc_group'].cat.codes
df_mcc.head()

MCC                                           mcc_name  mcc_group
0   742                                Ветеринарные услуги          6
1   763                   Сельскохозяйственные кооперативы          6
2   780          Услуги садоводства и ландшафтного дизайна          6
3  1520  Генеральные подрядчики – жилое и коммерческое ...          6
4  1711  Генеральные подрядчики по вентиляции, теплосна...          6

__17__. Присоедините к таблице `df` столбец `mcc_group`. Какой тип `JOIN`-а подойдет лучше всего для этой задачи? Результат сохраните в `df`.

In [135]:
df = pd.merge(df, df_mcc, how='left', on='MCC')


In [137]:
df = df.drop(['mcc_name'], axis=1)
df.iloc[index].head()

PERIOD  cl_id   MCC channel_type  currency TRDATETIME  amount  \
51305   01/06/2017    913  7230        type2       810 2017-06-14  2100.0   
226863  01/07/2017   4183  5499        type1       810 2017-07-20   140.0   
124479  01/11/2017   2253  5735        type2       810 2017-11-28   397.0   
11132   01/08/2017    225  5411        type4       810 2017-08-07    44.9   
473017  01/03/2017   9869  5812        type1       810 2017-03-30   350.0   

       trx_category  target_flag  target_sum  channel_type_label  \
51305           POS            1   449004.29                   1   
226863          POS            1   132726.55                   0   
124479          POS            1       59.00                   1   
11132           POS            1   466235.45                   3   
473017          POS            0        0.00                   0   

        trx_category_label  trx_category_mul  mday  year  ymonth  wday  \
51305                    6                -1    14  2017       6     2   
226863                   6                -1    20  2017       7     3   
124479                   6                -1    28  2017      11     1   
11132                    6                -1     7  2017       8     0   
473017                   6                -1    30  2017       3     3   

        is_weekend  mcc_group  
51305        False        7.0  
226863       False       17.0  
124479       False       15.0  
11132        False       17.0  
473017       False       15.0

__18__. Выведите сумму, среднее и среднеквадратичное отклонение трат `amount` по каждой из категорий `trx_category` в рамках канала привлечения `channel_type`.

In [147]:



df_group = (
    df
    .groupby([
        'trx_category', 
        'channel_type'
    ]) [['amount']]
    .sum()
)
(
    df_group
    .reset_index(inplace=True) #заменяет индекс на стандартный от 0 до N_lines - 1
)

df_group.head()


trx_category channel_type     amount
0     BACK_TRX        type1  2.059e+07
1     BACK_TRX        type2  3.422e+06
2     BACK_TRX        type3  3.222e+05
3     BACK_TRX        type4  2.259e+05
4     BACK_TRX        type5  6.224e+04

In [148]:

df_group = (
    df
    .groupby([
        'trx_category', 
        'channel_type'
    ]) [['amount']]
    .mean()
)
(
    df_group
    .reset_index(inplace=True) #заменяет индекс на стандартный от 0 до N_lines - 1
)

df_group.head(17)

trx_category channel_type     amount
0      BACK_TRX        type1  10973.767
1      BACK_TRX        type2   4700.521
2      BACK_TRX        type3  20139.201
3      BACK_TRX        type4   7531.287
4      BACK_TRX        type5   1682.229
5        C2C_IN        type1  13568.209
6        C2C_IN        type2  14020.438
7        C2C_IN        type3  10876.250
8        C2C_IN        type4  15291.915
9        C2C_IN        type5  10556.333
10      C2C_OUT        type1  16348.025
11      C2C_OUT        type2  15804.485
12      C2C_OUT        type3   5061.857
13      C2C_OUT        type4  20977.273
14      C2C_OUT        type5  13812.508
15     CASH_ADV        type1  25185.410
16     CASH_ADV        type2  32250.535

In [149]:

df_group = (
    df
    .groupby([
        'trx_category', 
        'channel_type'
    ]) [['amount']]
    .std()
)
(
    df_group
    .reset_index(inplace=True) #заменяет индекс на стандартный от 0 до N_lines - 1
)

df_group.head(17)

trx_category channel_type     amount
0      BACK_TRX        type1  55122.935
1      BACK_TRX        type2  12835.219
2      BACK_TRX        type3  35475.888
3      BACK_TRX        type4  28910.156
4      BACK_TRX        type5   2331.338
5        C2C_IN        type1  21544.663
6        C2C_IN        type2  25895.815
7        C2C_IN        type3  12332.637
8        C2C_IN        type4  12320.494
9        C2C_IN        type5  11777.934
10      C2C_OUT        type1  25596.879
11      C2C_OUT        type2  25665.438
12      C2C_OUT        type3   7569.332
13      C2C_OUT        type4  26067.515
14      C2C_OUT        type5  20574.017
15     CASH_ADV        type1  59272.348
16     CASH_ADV        type2  52204.208

__19__. Выведите топ-5 клиентов, с наибольшим оборотом по `С2С` транзакциям. Под оборотом подразумевается полный объем полученных и переведенных средств без учета знака (т.е. не имеет значения получал ли клиент средства на свой счет, или переводил их на другой счет).

In [205]:
df_c2c = (
    df
    .groupby([
        'trx_category', 
        'cl_id'
    ]) [['amount']]
    .sum()
)
(
    df_c2c
    .reset_index(inplace=True) #заменяет индекс на стандартный от 0 до N_lines - 1
)
df
df_i = df_c2c.loc[df_c2c['trx_category'] == 'C2C_IN', :]
df_o = df_c2c.loc[df_c2c['trx_category'] == 'C2C_OUT', :]
df_com = pd.merge(df_i, df_o, how='left', on='cl_id')
df_com['summ'] = df_com.loc['0':'4999', ['amount_x', 'amount_y']].sum(axis = 1)

df_com.nlargest(5, 'summ')

trx_category_x  cl_id   amount_x trx_category_y   amount_y       summ
2190         C2C_IN   4445  2.390e+06        C2C_OUT  8.474e+05  3.237e+06
704          C2C_IN   1408  6.000e+04        C2C_OUT  1.836e+06  1.896e+06
2074         C2C_IN   4211  7.866e+05        C2C_OUT  9.271e+05  1.714e+06
2646         C2C_IN   5373  1.262e+06        C2C_OUT        NaN  1.262e+06
2136         C2C_IN   4340  1.248e+06        C2C_OUT  1.040e+04  1.259e+06

__20__. Прологарифмируйте сумму трат `amount`, а результат сохраните в столбец `amount_log`. Как нужно логарифмировать, чтобы не было проблем при `amount == 0`? 

In [150]:
import math

df['amount_log'] = (
    df['amount']
    .apply(lambda x: math.log10(x) if x>0 else None)
)

__21__. Выполните onehot-кодирование для столбцов `channel_type` и `trx_category`. Полученные столбцы присоедините к таблице `df`, результат сохраните в переменную `df`.  

In [219]:
df_channel_type = pd.get_dummies(df['channel_type'])
df_channel_type.columns = ['channel_type_' + c for c in df_channel_type.columns]
df_channel_type.head()

df_trx_category = pd.get_dummies(df['trx_category'])
df_trx_category.columns = ['origin_' + c for c in df_trx_category.columns]
df_trx_category.head()
df = pd.concat([df, df_channel_type, df_trx_category], sort=False, axis=1)
df.iloc[index].head()

PERIOD  cl_id   MCC channel_type  currency TRDATETIME  amount  \
51305   01/06/2017    913  7230        type2       810 2017-06-14  2100.0   
226863  01/07/2017   4183  5499        type1       810 2017-07-20   140.0   
124479  01/11/2017   2253  5735        type2       810 2017-11-28   397.0   
11132   01/08/2017    225  5411        type4       810 2017-08-07    44.9   
473017  01/03/2017   9869  5812        type1       810 2017-03-30   350.0   

       trx_category  target_flag  target_sum  channel_type_label  \
51305           POS            1   449004.29                   1   
226863          POS            1   132726.55                   0   
124479          POS            1       59.00                   1   
11132           POS            1   466235.45                   3   
473017          POS            0        0.00                   0   

        trx_category_label  trx_category_mul  mday  year  ymonth  wday  \
51305                    6                -1    14  2017       6     2   
226863                   6                -1    20  2017       7     3   
124479                   6                -1    28  2017      11     1   
11132                    6                -1     7  2017       8     0   
473017                   6                -1    30  2017       3     3   

        is_weekend  mcc_group  amount_log  channel_type_type1  \
51305        False        7.0       3.322                   0   
226863       False       17.0       2.146                   1   
124479       False       15.0       2.599                   0   
11132        False       17.0       1.652                   0   
473017       False       15.0       2.544                   1   

        channel_type_type2  channel_type_type3  channel_type_type4  \
51305                    1                   0                   0   
226863                   0                   0                   0   
124479                   1                   0                   0   
11132                    0                   0                   1   
473017                   0                   0                   0   

        channel_type_type5  origin_BACK_TRX  origin_C2C_IN  origin_C2C_OUT  \
51305                    0                0              0               0   
226863                   0                0              0               0   
124479                   0                0              0               0   
11132                    0                0              0               0   
473017                   0                0              0               0   

        origin_CASH_ADV  origin_CAT  origin_DEPOSIT  origin_POS  \
51305                 0           0               0           1   
226863                0           0               0           1   
124479                0           0               0           1   
11132                 0           0               0           1   
473017                0           0               0           1   

        origin_WD_ATM_OTHER  origin_WD_ATM_PARTNER  origin_WD_ATM_ROS  \
51305                     0                      0                  0   
226863                    0                      0                  0   
124479                    0                      0                  0   
11132                     0                      0                  0   
473017                    0                      0                  0   

        channel_type_type1  channel_type_type2  channel_type_type3  \
51305                    0                   1                   0   
226863                   1                   0                   0   
124479                   0                   1                   0   
11132                    0                   0                   0   
473017                   1                   0                   0   

        channel_type_type4  channel_type_type5  origin_BACK_TRX  \
51305                    0                   0                0   
226863           

__22__. Сохраните итоговую подтаблицу в файл.

In [220]:
df.iloc[index].to_csv('answer.csv', index=False, sep=',')

## Бонусная часть

Попробуйте придумать решающее правило (без использования алгоритмов машинного обучения), по которому можно было бы предсказать значение переменной `target_flag`. **Запрещается** использовать переменную `target_sum`.

Разобъем выборку на тренировочную и отложенную.

In [35]:
from sklearn.model_selection import train_test_split

mask = df.columns.str.startswith('target_')

target_cols  = df.columns[mask]
feature_cols = df.columns[~mask]

X, y = df.loc[:, feature_cols], df.loc[:, target_cols]

X_train, X_test, y_train, y_test = \
    train_test_split(X, y, test_size=0.3, shuffle=True, random_state=9872)

Ваше решение должно побить следующий бейзлайн:

```python
from sklearn.metrics import accuracy_score

assert accuracy_score(y_test['target_flag'], y_pred) > 0.7
```

**Запрещается** использовать отложенную выборку `(X_test, y_test)` для выбора решения.

Что показывает метрика `accuracy`?

In [36]:
# put your answers here

### Вопросы и подсказки

Исследуйте рапределения и зависимости между признаками отностительно целевой переменной `target_flag`. Для решения этой задачи вам может быть полезна визуализация.

1. Какие любопытные выводы можно сделать из такого анализа? 
2. Какие проблемы возникают с визуализацией совместного распредления категориальных (дискретных) и непрерывных значений? Как эту проблему можно решить?
3. В исходном датасете очень много записей, поэтому визуализация может работать **неприемлемо** долго. Как решить эту проблему?

Приведите ниже код и комментарии с ответами на вопросы выше.

In [ ]:
# put your answers here

In [ ]:
# put your code here

### Решение

In [40]:
# put your code here

In [41]:
accuracy_score(y_test['target_flag'], y_pred)

0.7309009608981067